### DISCUSSION TITLE
TextMarkers/Segments and deletion

### DISCUSSION THREAD

**Discussion by DoctorVanGogh** [icsharpcode/AvalonEdit#327] [Q&A]
I've been playing around with the TextMarker example from https://github.com/siegfriedpammer/AvalonEditSamples and have encountered a problem when the text underpinning a marker is removed from the editor.

Basically I can capture a `TextSegment`'s `OnSegmentChanged` and use that to remove the segment if it has been reduced to zero length. This works fine if the text edit is performed from *within* the segment's bounds:
![in-segment](https://user-images.githubusercontent.com/324067/146383326-acf03a4a-3eec-4b01-a75e-6e72fb333ccb.gif)

**If however** the entire segment's contents are deleted (via a selection which encompasses the segment and more), a zero length marker will stay.:
![out-of-segment](https://user-images.githubusercontent.com/324067/146383446-15d4b993-34c5-42fa-a045-a9d823486a92.gif)

The segment disappears from the appropriate `TextSegmentCollection` (`markers` is empty in https://github.com/siegfriedpammer/AvalonEditSamples/blob/bd6858eed774743b92d971b179b3bd3bb32655e7/TextMarkerSample/SharpDevelop/TextMarkerService.cs#L95) only to reappear during a subsequent Redraw (`markers` is **no longer** empty in https://github.com/siegfriedpammer/AvalonEditSamples/blob/bd6858eed774743b92d971b179b3bd3bb32655e7/TextMarkerSample/SharpDevelop/TextMarkerService.cs#L123).

How do I catch the second case?

**Comment by siegfriedpammer** [comment]
As far as I can remember the `TextMarker` was not meant to be deletable by document edits. In our past use-cases (in https://github.com/icsharpcode/SharpDevelop, where I would recommend you to take a look, if you wonder how to use a certain feature of AvalonEdit...) it was always some external event (e.g. parser updates, refreshing the list of errors by re-running the build) that caused text markers to be removed from the view. The API that's meant to be used for use-cases like this is [`ITextAnchor`](https://github.com/icsharpcode/AvalonEdit/blob/master/ICSharpCode.AvalonEdit/Document/ITextAnchor.cs).

I won't comment on any of your suggestions in #331 because https://github.com/icsharpcode/AvalonEdit/blob/47f61e9427b0cb2653f83e8c303b9ea94c8ef60c/ICSharpCode.AvalonEdit/Document/TextSegmentCollection.cs#L55-L56 definitely does not apply to me ;-), so I will leave that to @dgrunwald.

Nonetheless, I hope this helps!

**Comment by dgrunwald** [comment]
SharpDevelop's `TextMarker` was meant for highlighting provided by external sources (e.g. compiler errors).
If you are trying to build a rich text editor, look into AvalonEdit's `RichTextModel`+`RichTextColorizer` classes instead.
